Select tokens

saralein opened the pull request:
Adds ability to select tokens to initializer.

zdavy commented on Rakefile:
```diff
@@ -7,5 +9,5 @@ end
 desc 'runs game'
 task :play do
-  ruby 'app.rb'
+  initializer(3)
```
💯

zdavy commented on app.rb:
```diff
@@ -6,20 +6,40 @@
+
+  user_interface.welcome
+  token1 = select_token(user_interface, setup_script, setup_validator, player1, '')
+  token2 = select_token(user_interface, setup_script, setup_validator, player2, token1)
```
I think it's probably worth breaking out a GameSetup class and a Configuration object at this point, instead of some helper functions. That way inside of this game initializer we don't have any real logic, all it needs to do is wire everything up, welcoming the user and setting up the tokens can happen elsewhere.

saralein replied:
Now have a game setup class with a configuration struct.

zdavy commented on app.rb:
```diff
@@ -6,20 +6,40 @@
 end
 
-initializer(3, 'X', 'O')
+def select_token(user_interface, script, setup_validator, player, compare_to)
```
could we make generating `"\n#{player}#{script.token_selection}"` a method on the script instead?

saralein replied:
Added this to the script.

damonkelley commented on lib/game.rb:
```diff
@@ -23,6 +24,6 @@ def take_turn
 
   def end_game
-    @current_player.end_game(@checker.winner)
+    @current_player.end_game(@checker.winner, @player1.token)
```
This works because the player 1 is always the human player, yeah?

It feels a little off that we would assign `player1` to two different instance variables.

saralein replied:
Removed this assignment and presented the winner in another fashion.

zdavy commented on lib/ui/user_interface.rb:
```diff
@@ -1,13 +1,15 @@
 
   def welcome
-    @io.display_message("\nWelcome to Tic Tac Toe. :D")
+    display_message("\nWelcome to Tic Tac Toe. :D")
```
I've grown pretty fond of those Script containers, can we move these strings over to a script?

saralein replied:
Should now be in the script.

saralein committed:
Adds ability for user to select token.

Adds ability to select user tokens.

Fixes arguments in UI in app.

Adds select token option to initializer with token validation.

Moves token selection into game setup class with a game script.

Moves some prompts into game script.

Adjusts game setup to clear and display board before selecting token.

Moves player script and validator into scripts and validators folder.

Adjusts order of methods inside game_setup.

Updates name of player validator class.

Updates end game method to not take additional parameter.

Updates create token prompt method on game script.

Adds more script to game script.

Adjusts test output for updates to game script.

Moves it block to correct describe in game setup test.

Adjusts setup validator.

Adjusts script with new lines in some prompts.

Adds game runner to pull logic out of initializer.

Moves lambda out of game runner. Calls initializer in game runner for replay/restart.

Removes extraneous scripts pieces. Adds new line to win prompt.

Adds new line to draw prompt.

Adds mode, order, errors to game script.

Removes instance variable moves and generates in display_board. Adjusts welcome method.

Generalizes setup validator for use in game setup.

Adds attribute accessor for game state.

Creates mock game for testing.

Adds reset method to board for access in game setup.

Modifies game runner to check game state. Adds UI to game runner init.

Moves config setup into game setup. Adds mode, order, board size options.

Adjusts new lines for prompts.

Adjusts expected output in ui tests.

Moves board_sizes, player_order, modes to class variables.

Removes structs and makes classes for game config and progress. Adjusts tests and reorganizes files.

saralein commented:
Created a game setup and game runner class to pull logic out of initializer.

zdavy commented on app.rb:
```diff
@@ -1,67 +1,50 @@
-  unless(replay.to_sym == :yes || replay.to_sym == :y)
-    state.stop_playing = true
-  end
+def create_game_from_pieces(game_pieces, game_options, game_progress)
```
The only place that this method is used is to bpassed into the start playing, it feels a little weird to me, is there a way embed this method in the game?

saralein replied:
OK. I modified this a bit based on our discussion earlier.  I got rid of passing the lambda into the game runner. The initializer will call game setup and game runner.  Game runner will call initializer when the player wants to restart or replay.  Does this seem a little better? If not, what sort of direction do you think would improve it?

zdavy commented on app.rb:
```diff
@@ -1,67 +1,44 @@
+  human = Human.new(user_interface)
+  ai = AI.new(board, checker, game_options.token2, game_options.token1)
+  player1 = Player.new(human_script, human, game_options.token1, user_interface, player_validator)
+  player2 = Player.new(ai_script, ai, game_options.token2, user_interface, player_validator)
```
so you're getting data back from game setup, and applying that data to create players in the initialize function here, why not just have the game setup return two players?

saralein commented:
Reworked this and moved logic into game setup, which returns a config for game and game runner.

zdavy commented on app.rb:
```diff
@@ -1,67 +1,30 @@
-  Struct.new('State', :stop_playing, :is_won, :restart)
-  state = Struct::State.new(false, false, false)
+Struct.new('Configuration', :game_progress, :board, :checker, :game_script, :user_interface, :player_validator, :player1, :player2)
+Struct.new('GameProgress', :stop_playing, :is_won, :restart)
```
I think it's worth making these structs first class objects, like classes in their own files. The ruby Struct::MyStruct pattern makes them easy to use anywhere, but then you have to be cognizant of run time order of operations. If you create classes in their own files you can just "require" them as needed and not worry about loading the app.rb before any files like game_setup that require these structs.

damonkelley replied:
I am with @zacharyjdavy on this.

Plus, `Struct` does not really fit into the language you have created for the domain.

saralein replied:
I removed these structs and made classes.  Let me know if you are OK with the configuration class.  I created a method to set players later, but it's a little against the spirit of some earlier comments.

zdavy commented on lib/game_setup.rb:
```diff
@@ -0,0 +1,112 @@
+    @order = :player1
+    @board_size = {'1' => 3, '2' => 4}
+    @modes = {'1' => :humanVShuman, '2' => :humanVScomputer, '3' => :computerVScomputer}
+    @orders = {'1' => :player1, '2' => :player2}
```
modes, orders, and board_sizes don't appear to fit the bill of "instance" variable. They feel permanent. Let's refactor that a little bit.

saralein replied:
I made these [class variables](https://github.com/saralein/tic-tac-toe_ruby/pull/15/files#diff-b1e3c91fecf2e775c58f0ba6d8acb4ecR2).  Debated placing them in the methods since they are really only used there, but I liked them more visible.  Let me know what you think.
